File: 20.有效的括号.py
#
# @lc app=leetcode.cn id=20 lang=python3
#
# [20] 有效的括号
#

# @lc code=start
class Solution:
    def isValid(self, s: str) -> bool:
        map = {
            '(':')',
            '[':']',
            '{':'}'
        }

        stack = []

        for char in s:
            if char not in map and len(stack) == 0:
                return False
            elif char not in map and map[stack.pop()] != char:
                return False
            elif char in map:
                stack.append(char)
        
        return len(stack) == 0



# @lc code=end

File: 20.有效的括号.java
import java.util.HashMap;
import java.util.Stack;

/*
 * @lc app=leetcode.cn id=20 lang=java
 *
 * [20] 有效的括号
 */

// @lc code=start
class Solution {
    public boolean isValid(String s) {

        HashMap<Character,Character> map = new HashMap<Character,Character>();
        map.put('(',')');
        map.put('{','}');
        map.put('[',']');

        Stack<Character> stack = new Stack<Character>();

        for(int i=0;i<s.length();i++){
            if(map.containsKey(s.charAt(i))){
                stack.push(map.get(s.charAt(i)));
            }else{
                if(stack.empty()||stack.pop() != s.charAt(i)){
                    return false;
                }

            }
        }

        return stack.empty();
    }
}
// @lc code=end

